scipy/special/_basic.py

```diff
@@ -2033,18 +2054,16 @@ def lpn(n, z):
            https://people.sc.fsu.edu/~jburkardt/f77_src/special_functions/special_functions.html
 
     """
-    if not (isscalar(n) and isscalar(z)):
-        raise ValueError("arguments must be scalars.")
     n = _nonneg_int_or_fail(n, 'n', strict=False)
-    if (n < 1):
-        n1 = 1
-    else:
-        n1 = n
-    if iscomplex(z):
-        pn, pd = _specfun.clpn(n1, z)
-    else:
-        pn, pd = _specfun.lpn(n1, z)
-    return pn[:(n+1)], pd[:(n+1)]
+
+    z = np.asarray(z)
+
+    pn = np.zeros((n + 1,) + z.shape, dtype=z.dtype)
+    pd = np.zeros_like(pn)
+    _lpn(z, out = (np.moveaxis(pn, 0, -1),
+        np.moveaxis(pd, 0, -1))) # new axes must be last for the ufunc
+
+    return pn, pd
```

Comment:
```suggestion
    pn = np.zeros(z.shape + (n + 1,), dtype=z.dtype)
    pd = np.zeros_like(pn)
    _lpn(z, out = (pn, pd))

    return pn, pd
```

Two things here. First we don't need `pn` and `pd` to be zeroed out, so using `empty` and `empty_like` here can shave off some overhead, particularly for `n` large in value and `z` large in shape.

Second, the use of `np.moveaxis` seems to introduce a lot of overhead. Here's an example

```
In [3]: %timeit lpn(1000, 0.2)
9.88 µs ± 8.31 ns per loop (mean ± std. dev. of 7 runs, 100,000 loops each)
```

here is the result after my suggestions are applied

```
In [3]: %timeit lpn(1000, 0.2)
6 µs ± 2.84 ns per loop (mean ± std. dev. of 7 runs, 100,000 loops each)
```

and since this is a scalar case, here is the result on `main`

```
In [2]: %timeit lpn(1000, 0.2)
3.87 µs ± 18.2 ns per loop (mean ± std. dev. of 7 runs, 100,000 loops each)
```

The comparison is less favorable for smaller `n`. 

It's not clear to me why we're adding this overhead to override the default behavior for a gufunc with this signature.

The elements of, say,  `lpn(10, np.linspace(0, 1, 100))` each having shape `(100, 11)` seems least surprising to me. I see this as a family of gufuncs, parametrized by `n`, which controls the extend to which we're expanding the input.

Reply:
The zeros was preexisting! Happy to change that to empty.

Yes, so the gufuncs need to have the newaxis of size `n + 1` as the last axis. It feels natural to return that as the first axis to be consistent with other special functions, like `sph_harm` which returns `(m + 1, n + 1, ...)`. We could either:

1) Drop the consistency, do as you've written.

2) Add a moveaxis after the gufunc is applied.

3) Special-case out scalars, so no moveaxis is applied in the scalar case.

4) Not worry about it?

Reply:
Cool, I saw that the zeros were pre-existing. 

`sph_harm` isn't a gufunc so I'm not sure that's inconsistent. The order `m` and degree `n` are array args like any other, and it just behaves like any other ufunc. I guess you're thinking about trying to make `sph_harm` a gufunc in a backwards compatible way, and don't want to start a convention for gufuncs that wouldn't allow that?

Reply:
Yes, exactly that. `sph_harm` is literally next on my list.

From a user perspective, I think it's natural to have these new axes as first. That way you can do `p[i, j]` and get the mth and nth associated Legendre polynomial, rather than `p[..., i, j]`. But I agree it then requires the moveaxis along with the gufunc call, and if put them last no moveaxis is required.

Reply:
An argument **for** putting the axes at the end is that it is then consistent with how we handle stacks of matrices, where the last two dimensions represent the matrix dimensions.

Whichever we choose, we should keep it as a consistent convention for special.

Reply:
Good point. Let’s keep the behavior like you have it. It’s what makes the most sense if I don’t think of these as gufuncs, just that they’re using gufuncs under the hood. I’ll see if I can find more ways to cut down overhead.

Reply:
I think just do the moveaxis stuff after the gufunc call and only if input is not scalar.

Reply:
I've thought about this more too, and I think we should leave it as is. Yes, if we do it after the call it saves time here, but then the next batch of operations will be slower because it doesn't have a contiguous last dimension. We're just not benchmarking that.

I also think we shouldn't worry about the scalar case so much. It's bit like benchmarking `np.sqrt`, which will be slower than `math.sqrt` on account of going through the ufunc machinery. We don't normally worry about this for scalars. We can ultimately make the scalar version of this available through `cython_special` which will be fast for scalars.

Reply:
What about

`out=(obvious thing) if z.ndim > 0 else (pn, pd)`

on mobile and in a hurry so not typing it all out.

Reply:
I've just checked that in, to preserve the speed for scalar case. I do think we should reconsider whether this is worth it in the future, and a `cython_special` version could resolve it.

Reply:
Modulo `isscalar` not doing what you want, having the conditional adds negligible overhead (single digit nanoseconds) in all cases to shave off several microseconds of overhead in the scalar case. The code isn't complicated unclear with the conditional. Why would this not be worth it?

Reply:
I've already made those changes, so what is just my opinion. To be honest, I don't think anyone will notice a few microseconds for a function like this. I think there is probably similar overhead in all the special functions that use ufuncs, but we haven't noticed as there weren't corresponding versions that directly called to Cython.